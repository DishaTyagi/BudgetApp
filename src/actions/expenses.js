import uuid from 'uuid';

//Action Generators for expenses.
//ADD_EXPENSE
export const addExpense = ( { description = '', note = '', amount = 0, createdAt = 0 } = {} ) => (      //toss export at the beginning for named exports.
    {       //action object
        type: 'ADD_EXPENSE',
        expense: {
            id: uuid(),
            //remaining 4 keys are user-defined.
            description,
            note,
            amount,
            createdAt
        }
    }
)

//REMOVE_EXPENSE
export const removeExpense = ( {id} = {} ) => (     //take id as obj prop only rather than passing it simply.
    {
        type: 'REMOVE_EXPENSE',
        id
    }
)

//EDIT_EXPENSE
export const editExpense = (id , updates) => (        //take it as an object or it will be problematic during jest testing
    {
        type: 'EDIT_EXPENSE',
        id,
        updates
    }
)

