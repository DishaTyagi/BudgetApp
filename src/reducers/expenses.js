//Expenses reducer
const expensesReducerDefaultState = [];
export default ( state = expensesReducerDefaultState , action) => {
    switch(action.type){
        case 'ADD_EXPENSE' :            
            return [...state, action.expense];       //spread operator(returns a new array)
        case 'REMOVE_EXPENSE' :
            return state.filter( ( {id} ) => id !== action.id)      //{id} of the individual state in filter across the state array of objects.
        case 'EDIT_EXPENSE':
            return state.map( (expense) => {
                if(expense.id === action.id){
                    return {
                        ...expense,
                        ...action.updates      //action.updates is also an object so hgrab it and override the values of expense with action.updates
                    }
                }
                else{
                    return expense;
                }
            } )
        default: 
            return state;
    }
};

// export default expensesReducer;