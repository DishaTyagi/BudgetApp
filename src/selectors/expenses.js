//Get visible expenses
//show expenses based on the filters
import moment from 'moment';

export default ( expenses, {text, sortBy, startDate, endDate} ) => {

    return expenses.filter( (expense) => {      //filter is the array function here.
        const createdAtMoment = moment(expense.createdAt);
        
        const textMatch = expense.description.toLowerCase().includes(text.toLowerCase()) ;              //textMatch is to check if the description of expense has the word that contains the text passed by the user.
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true ;        //day to check if its the same day or before the expense createdAt
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true ;  //false of conditional is true so that if there is np endDate then don't filter the expenses ont he basis of filters. Just return true i.e. all the filters

        return textMatch && startDateMatch && endDateMatch;     //if both are true, then the expense is included in the new array.
    }).sort( ( a, b ) => {            //sort is chained coz expenses have to be sorted based on date or amount.
        if(sortBy === 'date'){
            return a.createdAt < b.createdAt ? 1 : -1 ;     //if b expense has to be on the top, toss 1, else -1.
        }
        else if(sortBy === 'amount'){
            return a.amount < b.amount ? 1 : -1 ;
        }
    }) ;
    
}