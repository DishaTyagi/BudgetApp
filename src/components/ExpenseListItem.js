import React from 'react';
import {removeExpense} from '../actions/expenses';
import { connect } from 'react-redux';

const ExpenseListItem =  ({dispatch, id, description, amount, createdAt}) => (
    <div>
        <p>{description}</p>
        <p>{amount} - {createdAt} </p>
        <button onClick = { () => {
            dispatch(removeExpense(id));
        } } >Remove</button>
    </div>
)

export default connect()(ExpenseListItem);      //this makes dispatch available to the components mentioned here.