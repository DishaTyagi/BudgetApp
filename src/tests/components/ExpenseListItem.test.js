import ExpenseListItem from '../../components/ExpenseListItem';
import {shallow} from 'enzyme';
import moment from 'moment';
import React from 'react';
import expenses from '../fixtures/expenses'

test('should render ExpenseListItem', () => {
    const wrapper = shallow(<ExpenseListItem {...expenses[0] }/>);
    expect(wrapper).toMatchSnapshot();
});