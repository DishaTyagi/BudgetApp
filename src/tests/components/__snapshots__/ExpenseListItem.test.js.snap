// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseListItem 1`] = `
<div>
  <[object Object]
    to="edit/1"
  >
    <h3>
      copy
    </h3>
  </[object Object]>
  <p>
    300
     - 
    -345600000
     
  </p>
</div>
`;
